File: HashTables/Two Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        // Hash map to store index of numbers already encountered
        std::unordered_map<int, int> hash; // num : index
        
        for(int i = 0; i < nums.size(); i++){
            if(hash.count(target-nums[i]) != 0){
                result.push_back(i);
                result.push_back(hash[target-nums[i]]);
                break;
            }
            hash[nums[i]] = i;
        }
        return result;
    }
};

File: HashTables/Two Sum.rs
use std::collections::HashMap;

// O(n)
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.len() == 2 {
            return vec![0,1];
        }

        let mut num_map = HashMap::with_capacity(nums.len());

        for i in 0..nums.len() {
            match num_map.get(&(target - nums[i])) {
                Some(x) if *x != i => return vec![i as i32, *x as i32],
                _ => num_map.insert(nums[i], i),
            };
        };
        vec![0, 0]
    }
}


// O(2n)
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.len() == 2 {
            return vec![0,1];
        }

        let mut num_map = HashMap::with_capacity(nums.len());

        for i in 0..nums.len() {
            num_map.insert(nums[i], i);
        }

        for i in 0..nums.len() {
            match num_map.get(&(target - nums[i])) {
                Some(x) if *x != i => return vec![i as i32, *x as i32],
                _ => ()
            }
        };
        vec![0, 0]
    }
}